File: src/providers/lnbits.ts
import type { Invoice, LightningProvider } from "../types";

/**
 * LNbits — self-hosted or public instance Lightning provider.
 * Get key: seu-lnbits.com → API info → Invoice/read key
 */
export class LNbitsProvider implements LightningProvider {
  private apiKey: string;
  private baseUrl: string;

  constructor(apiKey: string, baseUrl = "https://legend.lnbits.com") {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  async createInvoice(amountSats: number): Promise<Invoice> {
    const res = await fetch(`${this.baseUrl}/api/v1/payments`, {
      method: "POST",
      headers: { "Content-Type": "application/json", "X-Api-Key": this.apiKey },
      body: JSON.stringify({ out: false, amount: amountSats, memo: "L402 API access" }),
    });
    if (!res.ok) {
      const err = await res.text().catch(() => res.statusText);
      throw new Error(`LNbits failed (${res.status}): ${err}`);
    }
    const data = (await res.json()) as { payment_request: string; payment_hash: string };
    const macaroon = Buffer.from(JSON.stringify({ hash: data.payment_hash, exp: Math.floor(Date.now() / 1000) + 3600 })).toString("base64");
    return { paymentRequest: data.payment_request, paymentHash: data.payment_hash, macaroon, amountSats, expiresAt: Math.floor(Date.now() / 1000) + 3600 };
  }

  async checkPayment(paymentHash: string): Promise<boolean> {
    const res = await fetch(`${this.baseUrl}/api/v1/payments/${paymentHash}`, { headers: { "X-Api-Key": this.apiKey } });
    if (!res.ok) return false;
    const data = (await res.json()) as { paid?: boolean };
    return !!data.paid;
  }
}

File: python/l402kit/providers/lnbits.py
import time, base64, json
import httpx
from ..types import Invoice


class LNbitsProvider:
    def __init__(self, api_key: str, base_url: str = "https://legend.lnbits.com"):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")

    async def create_invoice(self, amount_sats: int) -> Invoice:
        async with httpx.AsyncClient() as client:
            res = await client.post(
                f"{self.base_url}/api/v1/payments",
                headers={"X-Api-Key": self.api_key},
                json={"out": False, "amount": amount_sats, "memo": "L402 API access"},
            )
            res.raise_for_status()
            data = res.json()
            exp_ms = int(time.time() * 1000) + 3_600_000
            macaroon = base64.b64encode(json.dumps({"hash": data["payment_hash"], "exp": exp_ms}).encode()).decode()
            return Invoice(data["payment_request"], data["payment_hash"], macaroon, amount_sats, int(time.time()) + 3600)

    async def check_payment(self, payment_hash: str) -> bool:
        async with httpx.AsyncClient() as client:
            res = await client.get(f"{self.base_url}/api/v1/payments/{payment_hash}", headers={"X-Api-Key": self.api_key})
            return res.is_success and res.json().get("paid", False)
